File: mp7/DNA.py
"""
CS 1 Fall 2022 MP7
MP7 Provided DNA.py for reference in Porting Python to Java (students
do not need to modify this program in MP7).
Original Author: El Hovik

Defines a DNA class representing a single strand of DNA comprised
of the 4 base nucleotides, 'A', 'T', 'C', 'G'.
"""


class DNA:
    """
    Represents a DNA strand comprised of 4 base nucleotides. Each nucleotide
    is one of 'A', 'T', 'C', 'G' (standardized to uppercase).

    Attributes:
        - seq (str): Sequence string of DNA nucleotide bases.
    """
    # not required but a useful class constant
    BASES = {
        'A': 'T',
        'T': 'A',
        'G': 'C',
        'C': 'G'
    }

    def __init__(self, seq):
        """
        Initializes a new DNA sequence using the provided seq str.
        Case insensitivity is supported for the given seq, but a DNA sequence is
        standardized in upper-case.

        Arguments:
            - seq (str): sequence of 'A', 'T', 'C', 'G' bases.

        Raises:
            - ValueError if provided seq is not comprised of ATCG
              nucleotide bases.
        """
        seq = seq.upper()
        for base in seq:
            if base != 'A' and base != 'T' and base != 'C' and base != 'G':
                err_msg = 'Invalid DNA sequence. Must only contain ' + \
                          'ATCG bases.'
                raise ValueError(err_msg)
        self.seq = seq

    def __str__(self):
        """
        Returns the string representation of this DNA strand, which is simply
        the sequence of nucleotides.

        Returns:
            - (str): String representation of this DNA strand, e.g. 'ATACGC'
        """
        return self.seq

    def __len__(self):
        """
        Returns the length of the DNA, defined as the number of bases in
        its sequence.

        Returns:
            - (int): length of this DNA strand
        """
        return len(self.seq)

    def base_complement(self, base):
        """
        Returns the complement of the provided base character,
        ignoring letter-case (the complement is returned in upper-case).

        Arguments:
            - base (str): nucleotide base character

        Returns:
            - (str): complement of `base`, in upper-case

        Raises:
            - ValueError if the passed base is not a valid DNA nucleotide base.
        """
        base = base.upper()
        if base == 'A':
            return 'T'
        elif base == 'T':
            return 'A'
        elif base == 'C':
            return 'G'
        elif base == 'G':
            return 'C'
        else:
            raise ValueError('Invalid base.')

    def complement(self):
        """
        Returns the complement DNA sequence. For example, if this DNA
        is comprised of the bases 'ATACGC', its complement is returned as
        'TATGCG'.

        Returns:
            - (str): the complement DNA sequence.
        """
        result = ''
        for ch in self.seq:
            result += self.base_complement(ch)
        return result

    def count_occurrences(self, base):
        """
        Returns the number of occurrences of the given nucleotide base,
        ignoring letter-casing.

        Arguments:
            - base (str): nucleotide base char (one of ATCG)

        Returns:
            - (int): count of given base in this sequence

        Raises:
            - ValueError if given an invalid base
        """
        base = base.upper()
        if base != 'A' and base != 'T' and base != 'C' and base != 'G':
            raise ValueError('Invalid base.')
        return self.seq.count(base)

    def percentage_of(self, base):
        """
        Returns the percentage of the given base (ignoring
        letter-casing) in this DNA sequence as a float value
        between 0.0 and 1.0. (100%). Note that the percentage of any
        valid base in an empty DNA will be considered 0.0 (but the base
        validation will take precedence).

        Arguments:
            - base (str) nucleotide base char (one of ATCG)

        Returns:
            - (float): Percentage of base between 0.0 and 1.0

        Raises:
            - ValueError if given an invalid base
        """
        # self.count_occurrences does the error-handling for use, so don't
        # duplicate it here
        occurrences = self.count_occurrences(base)
        if occurrences == 0:
            return 0.0  # handles '' case too
        return occurrences / len(self.seq)

File: mp7/DNA.java
/**
 * CS 1 22fa MP 7 Parts B/C
 * Porting Classes in Python and Java
 * Student name: Lea Grohmann
 * Starter code for DNA.java
 * @author (original): El Hovik and Adam Abbas
 * 
 * This class represents a DNA sequence comprised of A', 'C', 'G', 'T' 
 * nucleotides. The class supports simple methods to represent the sequence 
 * as a String, get a complement sequence, compute information about
 * nucleotide counts and percentages, and transcription and translation to
 * mRNA and polypeptide Strings (chains of amino acids).
 */
import java.util.Random;

public class DNA {
    // String of DNA base characters in upper-case
    private String seq;
    // For the optional B.6. exercise
    private static Random mutator = new Random();
    private static final double MUTATION_RATE = 0.1;

    /**
     * Initializes a new (single-strand) DNA sequence using the provided 
     * seq String. A DNA sequence contains a sequence of the 4 DNA base 
     * nucleotides, 'A', 'T', 'C', 'G' (standardized to uppercase).
     * 
     * @param seq - sequence of 'A', 'T', 'C', 'G' bases.
     * @throws IllegalArgumentException if the sequence
     * is not comprised of ATCG nucleotide bases.
     */
    public DNA(String seq) {
        seq = seq.toUpperCase();
        for (int i = 0; i < seq.length(); i++) {
            if (seq.charAt(i) != 'A' && seq.charAt(i) != 'T' && seq.charAt(i) != 'C' && seq.charAt(i) != 'G') {

                String message = "Invalid DNA sequence. Must only contain ATCG bases.";
                throw new IllegalArgumentException(message);
            }
        }
        this.seq = seq;
    }

    /**
     * Returns the string representation of the DNA sequence,
     * defined as the sequence of nucleotide bases.
     * 
     * @return string sequence of nucleotide bases
     */
    public String toString() {
        return this.seq;
    }

    /**
     * Returns the size of the DNA sequence, defined as the number of 
     * nucleotide bases it contains.
     * 
     * @return - number of bases in the sequence
     */
    public int size() { 
        return this.seq.length();
    }

    /**
     * (Provided private helper method, do not change)
     * Returns the complement of the provided base character,
     * ignoring letter-case (the complement is returned in upper-case).
     * 
     * @param base - nucleotide base
     * 
     * @return - complement of `base`, in upper-case.
     * 
     * @throws IllegalArgumentException if the passed base
     * is not a valid DNA nucleotide base character.
     */
    private static char baseComplement(char base) {
        // to upper-case a char in Java, we need to use
        // the Character class.
        base = Character.toUpperCase(base);
        if (base == 'A') {
            return 'T';
        } else if (base == 'T') {
            return 'A';
        } else if (base == 'C') {
            return 'G';
        } else if (base == 'G') {
            return 'C';
        } else {
            String errMsg = "Invalid base.";
            throw new IllegalArgumentException(errMsg);
        }
    }

    /**
     * Returns the complement sequence of the DNA sequence
     * 
     * @return - complement DNA sequence
     */
    public String complement() {
        String result = "";

        for (int i = 0; i < seq.length(); i++) {
            result += baseComplement(seq.charAt(i));
        }
        return result;
    }

    /**
     * Returns the number of occurences of a base in the DNA sequence
     * 
     * @param base - nucleotide base
     * 
     * @return - number of occurances of base
     * 
     * @throws IllegalArgumentException if passed a base is not
     * a valid DNA nucleotide base character.
     */
    public int countOccurrences(char base) {
        int result = 0;
        base = Character.toUpperCase(base);

        if (base != 'A' && base != 'T' && base != 'C' && base != 'G') {
            throw new IllegalArgumentException("Invalid base.");
        }
        for (int i = 0; i < seq.length(); i++) {
            if (base == seq.charAt(i)) {
                result++;
            }
        }
        return result;
    }

    /**
     * Returns the percentage of a base present in the DNA sequence
     * 
     * @param base - nucleutide base
     * 
     * @return - percentage of base in DNA
     * 
     * @throws IllegalArgumentException if the base passed is not a
     * valid DNA nucleotide base character.
     */
    public double percentageOf(char base) {
        int occurances = countOccurrences(base);

        if (occurances == 0) {
            return 0.0;
        }

        return (double) occurances / size();
    }

    /**
     * Transcribes the current DNA sequence to mRNA, by 
     * taking the complement of all of the bases and 
     * switching the "T" nucleotide to "U".
     * 
     * @return the transcribed mRNA
     */
    public mRNA transcribe() { 
        String compDNA = complement();
        String seqRNA = compDNA.replace('T', 'U');
        return new mRNA(seqRNA);
    }

    /**
     * Translates the current DNA to an Amino Acid chain by 
     * transcribing to an RNA chain, finding the start codon 
     * ("AUG") and then parsing 3-letter codons and looking
     * up their respective amino acids in the "codon.txt" files.
     * 
     * @return full string of translated Amino Acids
     */
    public String translate() {
        String polypeptide = transcribe().toPolypeptide();
        return polypeptide;
    }

    /**
     * Optional B.6.
     */
    public DNA mutate(int mutationTime) {
        String mutatedSeq = this.seq;
        // Provided pseudocode:
        // For mutationTime times:
        // if mutator.nextDouble() < MUTATION_RATE
        // randomly select an index of the mutatedSeq, and
        // replace mutatedSeq with the prefix (unchanged), the complement
        // of the randomly-selected character, and the suffix (unchanged)
        // Note that the result String should have the same length as this.seq
        // and this.seq should not be changed as a result of this method.
        // TODO (Optional): B.6.

        return new DNA(mutatedSeq);
    }
}
